edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.AuthAgent/AuthAgentListener.cs

```diff
@@ -0,0 +1,532 @@
+            if (currentListener == null)
+            {
+                Events.ClosedWhenNotRunning();
+                return;
+            }
+
+            // this disposes the instance making the listener loop exit
+            Volatile.Write(ref exitListenerLoop, true);
+            currentListener.Close();
+
+            await this.listenerTask;
+            this.listenerTask = null;
+
+            Events.Closed();
+        }
+
+        public void Dispose() => this.CloseAsync(CancellationToken.None).Wait();
+
+        private void Start()
+        {
+            var newListener = new HttpListener();
+
+            if (Interlocked.CompareExchange(ref this.listener, newListener, null) != null)
+            {
+                newListener.Close();
+                Events.StartedWhenAlreadyRunning();
+                throw new InvalidOperationException("Cannot start AuthAgent twice");
+            }
+
+            Volatile.Write(ref this.exitListenerLoop, false);
+
+            SetupTimeouts(newListener);
+            this.listener.Prefixes.Add(this.listeningAddress);
+            this.listener.Start();
+
+            this.listenerTask = ListenerLoop();
+            this.listenerTask.ContinueWith(_ => Restart());
+        }
+
+        private void Restart()
+        {
+            // no need for restart, stopped on purpose
+            if (Volatile.Read(ref exitListenerLoop))
+            {
+                return;
+            }
+
+            var prevListener = Interlocked.Exchange(ref this.listener, null);
+            prevListener?.Close();
+
+            Events.Restarting();
+            Start();
+        }
+
+        private async Task ListenerLoop()
+        {
+            while (true)
+            {
+                try
+                {
+                    var context = await this.listener.GetContextAsync();
+
+                    Events.AuthRequestReceived();
+
+                    var requestString = await ReadRequestBodyAsStringAsync(context.Request);
+                    var requestRecord = Decode(requestString);
+
+                    await requestRecord.Match(
+                        async r => await this.SendResponseAsync(r, context),
+                        async () => await SendErrorAsync(context)
+                    );                   
+                }
+                catch (ObjectDisposedException e) when (e.ObjectName == typeof(HttpListener).FullName)
+                {
+                    Events.ShutdownListenerLoop();
+                    break;
+                }
+                catch (Exception e)
+                {
+                    Events.ErrorExitListenerLoop(e);
+                    break;
+                }
+            }
+        }
+
+        private async Task SendResponseAsync(AuthRequest request, HttpListenerContext context)
+        {
+            try
+            {
+                (var isAuthenticated, var credentials) =
+                    await GetIdsFromUsername(request)
```

Comment:
nit: this.

Reply:
if stylecop wasn't forcing it, its not in the project. You'll need to add it as a referance to the csproj. Its a simple copy-paste from the other projects. you might also want to sync with andrew since the dotnet 3.1 changes will need to happen for this new project as well.

Reply:
added code analysis/checkinbuild to the new project